### PR TITLE
likely_confounders

### PR DESCRIPTION
**Pull request by ghadialhajj**
Plot the distribution of misclassifications with respect to different values of likely confounders

### CODE REVIEW

File: immuneML/reports/ml_reports/ConfounderAnalysis.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+class ConfounderAnalysis(MLReport):
+    """
+    A report that plots the distributions of the false positives and the false negatives made by a given ML method in a barplot.
+    These metrics are ploted separately with respect to each of the metadata features specified by the user.
+
+    Arguments:
+
+        metadata_labels (list): A list of the metadata features to use as a basis for the calculations
+    """
```
Comment (BlueBasilisk) [inline immuneML/reports/ml_reports/ConfounderAnalysis.py:26]: could you add a YAML specification example? For example, see how it is done in the 'Coefficients' report

File: immuneML/reports/ml_reports/ConfounderAnalysis.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+from immuneML.ml_methods.MLMethod import MLMethod
+from immuneML.reports.ReportOutput import ReportOutput
+from immuneML.reports.ReportResult import ReportResult
+from immuneML.reports.ml_reports.MLReport import MLReport
+from immuneML.util.PathBuilder import PathBuilder
+
+
+class ConfounderAnalysis(MLReport):
+    """
+    A report that plots the distributions of the false positives and the false negatives made by a given ML method in a barplot.
```
Comment (BlueBasilisk) [inline immuneML/reports/ml_reports/ConfounderAnalysis.py:20]: From the description and the plotted result in the test, I am not really able to understand what the report does. Could you try to explain it a bit clearer?

for example (not sure if this is a correct description of what your report does):
This report plots the distributions of the given metadata labels for the false positive and false negative predicted examples in the test sets.

**Comment by pavlovicmilena** [comment]
@ghadialhajj could you make sure that the tests are passing? It seems that the test for the report you added is failing. It seems that the problem is the label name (signal_disease) which was recently updated not to include the "signal" prefix, but rather only what the user specifies in the YAML (which would be "disease" in this case). Could you maybe check this? :)

File: test/reports/ml_reports/test_counfounderAnalysis.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        report.ml_details_path = path / "ml_details.yaml"
+        report.label = "signal_disease"
+        report.result_path = path
+        report.train_dataset = self._encode_dataset(self._make_dataset(path / "train", size=100), path)
+        report.test_dataset = self._encode_dataset(self._make_dataset(path / "test", size=40), path, learn_model=False)
+        report.method = self._create_dummy_lr_model(path, report.train_dataset.encoded_data, report.label)
+
+        return report
+
+    def test_generate(self):
```
Comment (BlueBasilisk) [inline test/reports/ml_reports/test_counfounderAnalysis.py:99]: Some things I didn't understand about the output/test:
- why are age and HLA a boolean value?
- if you have 'false positives' and 'false negatives' on the y axis, why is this a score from -1 to 1? I was instead expecting a number (count) or value between 0-100/0-1 (percentage). 
- at least in the current example, no data seems to be plotted

File: immuneML/reports/ml_reports/ConfounderAnalysis.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+                plot.update_yaxes(title_text=f"{metric}", row=label_index + 1,
+                                  col=metric_index + 1)
+
+        plot.update_traces(marker_color=px.colors.sequential.Teal[3], showlegend=False)
+        filename = self.result_path / f"{output_name}.html"
+        plot.write_html(str(filename))
+        report_output_fig = ReportOutput(filename)
+        paths.append(report_output_fig)
+
```
Comment (BlueBasilisk) [inline immuneML/reports/ml_reports/ConfounderAnalysis.py:79]: it would be nice if in addition to returning a HTML file, the raw data could be exported as a .csv file (see for example how it is done in the 'Coefficients' report)

File: test/reports/ml_reports/test_counfounderAnalysis.py (hunk 3 above)
Comment (ghadialhajj) [reply test/reports/ml_reports/test_counfounderAnalysis.py:99]: Fixed the y-axis to be non-negative to avoid getting negative values on the y-axis when there are no misclassifications. It also only ticks at integer values.
Reduced the training set size for misclassifications to show up, to avoid empty plots.

File: immuneML/reports/ml_reports/ConfounderAnalysis.py (hunk 4 above)
Comment (ghadialhajj) [reply immuneML/reports/ml_reports/ConfounderAnalysis.py:103]: Now the report exports one .csv file for each metadata feature.

**Comment by BlueBasilisk** [comment]
I ran the tests for this branch locally and they're all passing for me, so I think it can be merged in?

**Comment by pavlovicmilena** [comment]
this is the branch where I added get_metadata for other dataset types, just need to add tests as well before merging [hopefully today] :) but yes, Ghadi has finished the report 🎉

**Comment by ghadialhajj** [comment]
Great :tada: :muscle: ! thank you for your help. Please let me know if there is anything else I can do :+1:

**Review by pavlovicmilena — APPROVED** [review]
